### PR TITLE
improved spock parameter detection

### PR DESCRIPTION
**Pull request by jonnybbb**

### CODE REVIEW

**Comment by jkschneider** [comment]
Thanks @jonnybbb!

**Comment by breskeby** [comment]
@jonnybbb did you test this already with the GE build?

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java [hunk 1]
```diff
@@ -36,6 +37,11 @@
 public class SpockParameterClassVisitor extends ClassVisitor {
 
     private static final Set<Character> REGEX_CHARS = setOf(new char[]{'<', '(', '[', '{', '\\', '^', '-', '=', '$', '!', '|', ']', '}', ')', '?', '*', '+', '.', '>'});
+    //A valid Java identifier https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8 including methods
```
Comment (JLLeitschuh) [inline plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]:
```suggestion
// A valid Java identifier https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8 including methods
```

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java [hunk 2]
```diff
@@ -58,10 +64,11 @@ public MethodVisitor visitMethod(int access, String name, String descriptor, Str
     public void visitEnd() {
         spockMethodVisitor.getTestMethodPatterns().stream()
             .filter(methodPattern -> {
-                String methodPatternRegex = escapeRegEx(methodPattern).replaceAll("#\\w+", "\\\\w+");
+                //detects a valid spock parameter and replace it with a wildcards http://spockframework.org/spock/docs/1.3/data_driven_testing.html#_more_on_unrolled_method_names
```
Comment (JLLeitschuh) [inline plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]:
```suggestion
// detects a valid spock parameter and replace it with a wildcards http://spockframework.org/spock/docs/1.3/data_driven_testing.html#_more_on_unrolled_method_names
```

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java [hunk 3]
```diff
@@ -36,6 +37,11 @@
 public class SpockParameterClassVisitor extends ClassVisitor {
 
     private static final Set<Character> REGEX_CHARS = setOf(new char[]{'<', '(', '[', '{', '\\', '^', '-', '=', '$', '!', '|', ']', '}', ')', '?', '*', '+', '.', '>'});
+    //A valid Java identifier https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8 including methods
+    public static final String SPOCK_PARAM_PATTERN = "#[\\p{L}\\d$_.()&&[^#\\s]]+";
+    public static final String PARAM_PLACEHOLDER = "#param";
+    public static final String WILDCARD_SUFFIX = ".*";
+    public static final String WILDCARD = WILDCARD_SUFFIX;
```
Comment (guylabs) [inline plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]: Why is that constant required, as it's the same value as `WILDCARD_SUFFIX`?

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java [hunk 4]
```diff
@@ -58,10 +64,11 @@ public MethodVisitor visitMethod(int access, String name, String descriptor, Str
     public void visitEnd() {
         spockMethodVisitor.getTestMethodPatterns().stream()
             .filter(methodPattern -> {
-                String methodPatternRegex = escapeRegEx(methodPattern).replaceAll("#\\w+", "\\\\w+");
+                //detects a valid spock parameter and replace it with a wildcards http://spockframework.org/spock/docs/1.3/data_driven_testing.html#_more_on_unrolled_method_names
+                String methodPatternRegex = escapeRegEx(methodPattern.replaceAll(SPOCK_PARAM_PATTERN, PARAM_PLACEHOLDER)).replaceAll(PARAM_PLACEHOLDER, WILDCARD) + WILDCARD_SUFFIX;
```
Comment (guylabs) [inline plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]: I'd extract the `methodPattern.replaceAll(SPOCK_PARAM_PATTERN, PARAM_PLACEHOLDER)` part to a `normalizeMethodName()` method to make it more readable.

File: plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java (hunk 3 above)
Comment (jonnybbb) [reply plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]: just for better readability

Comment (guylabs) [reply plugin/src/main/java/org/gradle/testretry/internal/visitors/SpockParameterClassVisitor.java]: I checked twice when reading the regex replacement ;). But lets keep it as it is.

**Pull request merged**